### PR TITLE
Check the number of parameters in `ParametrizedEvolution`

### PR DESCRIPTION
**Pull request by dwierichs**
This PR introduces a simple check for `ParametrizedEvolution` that tests whether the number of input parameters and the number of callable coefficients of the generating Hamiltonian matches.
An exception is made for `HardwareHamiltonian` generators, because they have a reordering function that meddles with the number of parameters.

### CODE REVIEW

**Review by Qottmann — APPROVED** [review]
looks good to me, thanks @dwierichs 👍 

approving pending the failing tests in `tests/ops/functions/test_equal.py` and  `tests/ops/functions/test_evolve.py`

File: tests/pulse/test_parametrized_evolution.py [hunk 1]
```diff
@@ -280,6 +283,28 @@ def test_hash_with_data(self):
         assert compare_to.hash != diff_ret_intmdt.hash
         assert compare_to.hash != diff_complementary.hash
 
+    def test_raises_wrong_number_of_params(self):
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply tests/pulse/test_parametrized_evolution.py:286]: sorry, I'm not entirely sure I understand what test you'd like to be added. 
1. One that checks that initializing a `ParametrizedHamiltonian` with the wrong number of parameters raises
2. One that checks with another instance of `ParametrizedHamiltonian` in `evolve`

[the remark this answers is not in the document]
Comment (mudit2812) [reply tests/pulse/test_parametrized_evolution.py:286]: Both `ParametrizedEvolution.__init__` and `ParametrizedHamiltonian.__call__` have checks to verify that the number of parameters. This function only checks that `ParametrizedEvolution` raises an error, but there is no test that explicitly tests whether `ParametrizedHamiltonian.__call__` raises the correct error as well.

**Review by mudit2812 — APPROVED** [review]
Thanks David!

**Pull request merged**